### PR TITLE
Add nni_auto_tune example

### PR DESCRIPTION
**Pull request by smallv0221**
Add nni_auto_tune examples for sift and glove100.

### CODE REVIEW

File: tools/nni-auto-tune/search_space.json [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+    "Samples": {"_type": "choice", "_value": [1000, 2000, 4000, 8000, 10000]},
+    "TPTNumber": {"_type": "choice", "_value": [32, 34, 38, 46, 62, 94, 128]},
+    "RefineIterations": {"_type": "choice", "_value": [2, 3]},
+    "NeighborhoodSize": {"_type": "choice", "_value": [16, 18, 22, 30, 46, 64]},
+    "CEF": {"_type": "choice", "_value": [1000, 1100, 1200, 1400, 1800,2000]},
+    "MaxCheckForRefineGraph": {"_type": "choice", "_value": [4096, 5120, 7168, 11264,16324]},
+    "NumberOfInitialDynamicPivots": {"_type": "choice", "_value": [1, 2, 4, 8, 16,32,50]},
+    "NumberOfOtherDynamicPivots": {"_type": "choice", "_value": [1, 2, 4, 8, 10]}
+}
+
```
Comment (chicm) [inline tools/nni-auto-tune/search_space.json:12]: NNI tuners such as TPE, SMAC, hyperband treat each value of type "choice" not related,  the order of the integers is not leveraged by NNI.  In order to make NNI search parameters more efficiently,  suggest to change parameter type from `choice` to `quniform` or `qloguniform`.
For example,  maybe we can change "CEF"  to :
```
"CEF": {"_type": "quniform", "_value": [1000, 2000, 100]}
```
NNI search spaces: https://nni.readthedocs.io/en/stable/hpo/search_space.html#randint

Comment (chicm-ms) [inline tools/nni-auto-tune/search_space.json:12]: NNI tuners such as TPE, SMAC, hyperband treat each value of type "choice" not related,  the order of the integers is not leveraged by NNI.  In order to make NNI search parameters more efficiently,  suggest to change parameter type from `choice` to `quniform` or `qloguniform`.
For example,  maybe we can change "CEF"  to :
```
"CEF": {"_type": "quniform", "_value": [1000, 2000, 100]}
```
NNI search spaces: https://nni.readthedocs.io/en/stable/hpo/search_space.html#randint

Comment (smallv0221) [reply tools/nni-auto-tune/search_space.json:12]: Done

File: tools/nni-auto-tune/config_aml.yml [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  name: TPE
+
+  classArgs:
+
+    optimize_mode: maximize
+
+   
+
+trainingService:
+
```
Comment (suiguoxin) [inline tools/nni-auto-tune/config_aml.yml:38]: Pls remove these empty lines

File: tools/nni-auto-tune/main.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
+# Copyright (c) Microsoft Corporation
+# All rights reserved.
+#
+# MIT License
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
```
Comment (suiguoxin) [inline tools/nni-auto-tune/main.py:6]: Seems that the license header is outdated, please refer to [here](https://github.com/smallv0221/SPTAG/blob/add-ann-auto-tune/setup.py)

Comment (smallv0221) [reply tools/nni-auto-tune/main.py]: Done.

File: tools/nni-auto-tune/config_aml.yml (hunk 2 above)
Comment (smallv0221) [reply tools/nni-auto-tune/config_aml.yml]: Done.

**Review by suiguoxin — APPROVED** [review]

File: tools/nni-auto-tune/plot.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+                             res["build_params"], res["search_params"]))
+            except:
+                print(os.path.join("results", path, p))
+
+    data.sort(key=lambda t: (-1 * t[0], -1 * t[1]))
+    last_x = -1
+    comparator = ((lambda xv, lx: xv > lx) if last_x < 0 else
+                  (lambda xv, lx: xv < lx))
+    for i, (q, r, b, bp, sp) in enumerate(data):
+        if comparator(r, last_x):
```
Comment (suiguoxin) [inline tools/nni-auto-tune/plot.py:37]: Can you explain the selection rule here?

**Comment by suiguoxin** [comment]
How do you get the performance of `hnswlib` and `scann`?

**Review by MaggieQi — APPROVED** [review]

File: tools/nni-auto-tune/model.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+class Sptag:
+
+    def __init__(self, algo, metric):
+        self._algo = str(algo)
+        self._para = {}
+        self._metric = {'angular': 'Cosine', 'euclidean': 'L2'}[metric]
+
+    def fit(self, X, para=None):
+        self._sptag = SPTAG.AnnIndex(self._algo, 'Float', X.shape[1])
```
Comment (suiguoxin) [inline tools/nni-auto-tune/model.py:51]: what does the parameter 'Float' mean here?

File: tools/nni-auto-tune/main.py [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+        label = data_reader.label
+    else:
+        X_train = DataReader(args.train_file,
+                             args.dim,
+                             batchsize=-1,
+                             datatype=args.data_type).readbatch()[1]
+        X_test = DataReader(args.query_file,
+                            args.dim,
+                            batchsize=-1,
+                            datatype=args.data_type).readbatch()[1]
```
Comment (suiguoxin) [inline tools/nni-auto-tune/main.py:134]: `X_train`, `X_test` might be type float64, which is not supported by SPTAG.
Please add something like this:
```
if X_train.dtype == '<f8' or X_test.dtype == '<f8':
    X_train = X_train.astype('float32')
    X_test = X_test.astype('float32')
```

File: tools/nni-auto-tune/main.py [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+        label = []
+        if args.label_file is None:
+            # if the groundtruth is not provided
+            # we calculate groundtruth distances with brute force
+            bf = BruteForceBLAS(distance)
+            bf.fit(X_train)
+            for i, x in enumerate(X_test):
+                if i % 1000 == 0:
+                    print('%d/%d...' % (i, len(X_test)))
+                res = list(bf.query_with_distances(x, 100))
```
Comment (suiguoxin) [inline tools/nni-auto-tune/main.py:146]: Why 100 here? `args.k` should be enough ?

File: tools/nni-auto-tune/main.py [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+                            args.dim,
+                            batchsize=-1,
+                            datatype=args.data_type).readbatch()[1]
+        distance = args.distance
+        dimension = args.dim
+        label = []
+        if args.label_file is None:
+            # if the groundtruth is not provided
+            # we calculate groundtruth distances with brute force
+            bf = BruteForceBLAS(distance)
```
Comment (suiguoxin) [inline tools/nni-auto-tune/main.py:141]: Need to specify the second construction parameter, `precision` here

**Review by suiguoxin — CHANGES_REQUESTED** [review]
Please resolve data type issues for int8 & float64

File: tools/nni-auto-tune/main.py [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+        data_reader = HDF5Reader(args.train_file)
+        X_train, X_test = data_reader.readallbatches()
+        distance = data_reader.distance
+        dimension = data_reader.featuredim
+        label = data_reader.label
+    else:
+        X_train = DataReader(args.train_file,
+                             args.dim,
+                             batchsize=-1,
+                             datatype=args.data_type).readbatch()[1]
```
Comment (suiguoxin) [inline tools/nni-auto-tune/main.py:130]: Need to assign `targettype` here

File: tools/nni-auto-tune/model.py [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+        self._para = {}
+        self._metric = {'angular': 'Cosine', 'euclidean': 'L2'}[metric]
+
+    def fit(self, X, para=None, data_type='float32'):
+        self._data_type = {
+            'float32': 'Float',
+            'int8': 'Int8',
+            'int16': 'Int16'
+        }[data_type]
+        self._sptag = SPTAG.AnnIndex(self._algo, self._data_type, X.shape[1])
```
Comment (suiguoxin) [inline tools/nni-auto-tune/model.py:56]: Does capitalization of `data_type` matter?

**Review by suiguoxin — APPROVED** [review]

**Review by MaggieQi — APPROVED** [review]

**Comment by smallv0221** [comment]
> @microsoft-github-policy-service agree [company="{Microsoft}"]

**Review by MaggieQi — DISMISSED** [review]
build ci

**Review by MaggieQi — APPROVED** [review]

**Pull request merged**